File: src/jaslib/jaslib/note/jpnote.py
from __future__ import annotations

from collections import defaultdict
from typing import TYPE_CHECKING, override

from autoslot import Slots
from jaslib.note.jpnote_data import JPNoteData
from jaslib.note.note_constants import MyNoteFields
from jaslib.note.note_flush_guard import NoteRecursiveFlushGuard
from jaslib.note.note_tags import NoteTags
from jaslib.note.tags import Tags
from jaslib.sysutils import ex_assert, ex_str
from jaslib.sysutils.weak_ref import WeakRef, WeakRefable
from typed_linq_collections.collections.q_set import QSet

from jaslib import app

if TYPE_CHECKING:
    from jaslib.note.collection.jp_collection import JPCollection

type JPNoteId = int

class JPNote(WeakRefable, Slots):
    def __init__(self, data: JPNoteData | None = None) -> None:
        self.weakref: WeakRef[JPNote] = WeakRef(self)
        self.recursive_flush_guard: NoteRecursiveFlushGuard = NoteRecursiveFlushGuard(self.weakref)
        self.__hash_value: int = 0

        self.tags: NoteTags = NoteTags(self.weakref, data)

        self._fields: dict[str, str] = data.fields if data else defaultdict(str)
        self._id_cache: JPNoteId = data.id if data else 0

    # noinspection PyUnusedFunction
    @property
    def is_flushing(self) -> bool: return self.recursive_flush_guard.is_flushing

    def get_data(self) -> JPNoteData: return JPNoteData(self.get_id(), self._fields, self.tags.to_interned_string_list())

    @override
    def __eq__(self, other: object) -> bool:
        ex_assert.not_none(self.get_id(), "You cannot compare or hash a note that has not been saved yet since it has no id")
        return isinstance(other, JPNote) and other.get_id() == self.get_id()

    def _update_in_cache(self) -> None: raise NotImplementedError()

    def is_studying(self, card_type: str | None = None) -> bool: raise NotImplementedError()  # pyright: ignore
    def is_studying_read(self) -> bool: raise NotImplementedError()
    def is_studying_listening(self) -> bool: raise NotImplementedError()


    @override
    def __hash__(self) -> int:
        if not self.__hash_value:
            self.__hash_value = int(self.get_id())
            ex_assert.that(self.__hash_value != 0, "You cannot compare or hash a note that has not been saved yet since it has no id")
        return self.__hash_value

    @override
    def __repr__(self) -> str:
        return f"""{self.get_question()}: {self.get_answer()}"""

    @property
    def collection(self) -> JPCollection:
        return app.col()

    def get_question(self) -> str:
        value = self.get_field(MyNoteFields.question)
        return value if value else "[EMPTY]"

    def get_answer(self) -> str:
        return self.get_field(MyNoteFields.answer)


    def get_direct_dependencies(self) -> QSet[JPNote]:
        return QSet()

    def _on_tags_updated(self) -> None:
        """Called when tags are modified. Subclasses can override to invalidate cached state."""
        pass

    def _get_dependencies_recursive(self, found: QSet[JPNote]) -> QSet[JPNote]:
        if self in found:
            return found
        found.add(self)
        for dependency in self.get_direct_dependencies():
            dependency._get_dependencies_recursive(found)
        return found

    # noinspection PyUnusedFunction
    def get_dependencies_recursive(self) -> QSet[JPNote]:
        return self._get_dependencies_recursive(QSet())

    def get_id(self) -> JPNoteId:
        return self._id_cache

    def set_id(self, id: JPNoteId) -> None:
        if self._id_cache != 0: raise RuntimeError("Cannot change id of a note that has already been saved")
        self._id_cache = id


    def update_generated_data(self) -> None:
        pass

    def get_field(self, field_name: str) -> str:
        return self._fields[field_name]

    def _is_persisted(self) -> bool:
        return self._id_cache != 0
        #raise NotImplementedError()
        #return int(self.backend_note.id) != 0

    def _flush(self) -> None:
        if self._is_persisted():
            self.recursive_flush_guard.flush()

    def set_field(self, field_name: str, value: str) -> None:
        field_value = self._fields[field_name]
        if field_value != value:
            self._fields[field_name] = value
            self._flush()

    # noinspection PyUnusedFunction
    def priority_tag_value(self) -> int:
        for tag in self.tags:
            if tag.name.startswith(Tags.priority_folder):
                return int(ex_str.first_number(tag.name))
        return 0

    def get_meta_tags(self) -> QSet[str]:
        tags: QSet[str] = QSet()
        for tag in self.tags:
            if tag.name.startswith(Tags.priority_folder):
                if "high" in tag.name: tags.add("high_priority")
                if "low" in tag.name: tags.add("low_priority")

        if self.tags.contains(Tags.TTSAudio): tags.add("tts_audio")

        return tags

    # noinspection PyUnusedFunction
    def get_source_tag(self) -> str:
        source_tags = [t for t in self.tags if t.name.startswith(Tags.Source.folder)]
        if source_tags:
            source_tags = sorted(source_tags, key=lambda tag: len(tag.name))
            return source_tags[0].name[len(Tags.Source.folder):]
        return ""

File: src/src_dotnet/JAStudio.Core/Note/JPNote.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.Note.Collection;
using JAStudio.Core.Note.CorpusData;
using JAStudio.Core.Note.NoteFields;
using MemoryPack;

namespace JAStudio.Core.Note;

public abstract class JPNote
{
   public NoteFlushGuard RecursiveFlushGuard { get; }
   public NoteGuard Guard { get; }
   public NoteServices Services { get; }
   bool _persisted;
   byte[]? _lastPersistedSnapshot;

   // Dictionary tracks card suspend status: true = active/unsuspended, false = suspended
   readonly Dictionary<string, bool> _cardStatus = new();

   public void CopyStudyingStatusFrom(JPNote other)
   {
      foreach(var kvp in other._cardStatus)
      {
         _cardStatus[kvp.Key] = kvp.Value;
      }
   }

   public NoteTags Tags { get; }

   readonly NoteId _id;

   protected JPNote(NoteServices services, NoteId id, List<string>? tags = null)
   {
      Services = services;
      RecursiveFlushGuard = new NoteFlushGuard(this);
      Guard = new NoteGuard(Flush);
      Tags = new NoteTags(this, tags);
      _id = id;
   }

   public bool IsFlushing => RecursiveFlushGuard.IsFlushing;

   public void SetStudyingStatus(CardStudyingStatus status)
   {
      // Update card status: true = active/unsuspended, false = suspended
      _cardStatus[status.CardType] = !status.IsSuspended;
   }

   public NoteData GetData() => ToCorpusData().ToNoteData();

   public override bool Equals(object? obj) => obj is JPNote other && other.GetId() == _id;

   public abstract void UpdateInCache();

   public bool IsStudying(string? cardType = null)
   {
      if(cardType == null)
      {
         // Return true if ANY card is active/unsuspended
         return _cardStatus.Any(kvp => kvp.Value);
      }

      // Return true if specific card type is active/unsuspended
      return _cardStatus.TryGetValue(cardType, out var isActive) && isActive;
   }

   public bool IsStudyingRead() => IsStudying(CardTypes.Reading);
   public bool IsStudyingListening() => IsStudying(CardTypes.Listening);

   public void SuspendAllCards()
   {
      Services.CardOperations.SuspendAllCardsForNote(_id);

      // Update local status for all known card types
      var cardTypes = _cardStatus.Keys.ToList();
      foreach(var cardType in cardTypes)
      {
         _cardStatus[cardType] = false; // false = suspended
      }

      Flush();
   }

   public void UnsuspendAllCards()
   {
      Services.CardOperations.UnsuspendAllCardsForNote(_id);

      // Update local status for all known card types
      var cardTypes = _cardStatus.Keys.ToList();
      foreach(var cardType in cardTypes)
      {
         _cardStatus[cardType] = true; // true = active/unsuspended
      }

      Flush();
   }

   public bool HasSuspendedCards()
   {
      // Return true if ANY card is suspended (value == false)
      return _cardStatus.Any(kvp => !kvp.Value);
   }

   public bool HasActiveCards() => IsStudying();

   public override int GetHashCode() => _id.GetHashCode();

   public override string ToString() => $"{GetQuestion()}: {GetAnswer()}";

   public JPCollection Collection => Services.Collection;

   public abstract string GetQuestion();

   public abstract string GetAnswer();

   public virtual void OnTagsUpdated()
   {
      // Called when tags are modified. Subclasses can override to invalidate cached state.
   }

   public NoteId GetId() => _id;

   public virtual void UpdateGeneratedData()
   {
      // Override in subclasses
   }

   public virtual List<MediaReference> GetMediaReferences() => [];

   public abstract CorpusDataBase ToCorpusData();

   byte[] TakeSnapshot() => MemoryPackSerializer.Serialize(ToCorpusData());

   internal void MarkAsPersisted()
   {
      _persisted = true;
      _lastPersistedSnapshot = TakeSnapshot();
   }

   public void Flush()
   {
      if(!_persisted) return;

      var currentSnapshot = TakeSnapshot();
      if(_lastPersistedSnapshot != null && currentSnapshot.AsSpan().SequenceEqual(_lastPersistedSnapshot))
         return;

      _lastPersistedSnapshot = currentSnapshot;
      RecursiveFlushGuard.Flush();
   }

   public int PriorityTagValue()
   {
      foreach(var tag in Tags)
      {
         if(tag.Name.StartsWith(Note.Tags.PriorityFolder))
         {
            var numberStr = StringExtensions.FirstNumber(tag.Name);
            if(int.TryParse(numberStr, out var num))
            {
               return num;
            }
         }
      }

      return 0;
   }

   public HashSet<string> GetMetaTags()
   {
      var tags = new HashSet<string>();

      foreach(var tag in Tags)
      {
         if(tag.Name.StartsWith(Note.Tags.PriorityFolder))
         {
            if(tag.Name.Contains("high")) tags.Add("high_priority");
            if(tag.Name.Contains("low")) tags.Add("low_priority");
         }
      }

      if(IsStudying(CardTypes.Reading) || IsStudying(CardTypes.Listening))
      {
         tags.Add("is_studying");
      }

      if(IsStudying(CardTypes.Reading)) tags.Add("is_studying_reading");
      if(IsStudying(CardTypes.Listening)) tags.Add("is_studying_listening");
      if(Tags.Contains(Note.Tags.TTSAudio)) tags.Add("tts_audio");

      return tags;
   }

   public string GetSourceTag()
   {
      var sourceTags = Tags.Where(t => t.Name.StartsWith(Note.Tags.Source.Folder)).ToList();
      if(sourceTags.Any())
      {
         var sorted = sourceTags.OrderBy(t => t.Name.Length).ToList();
         return sorted[0].Name.Substring(Note.Tags.Source.Folder.Length);
      }

      return string.Empty;
   }
}
